### ISSUE TITLE
Type implementing interface gives a union of fragment types instead of intersection

### ISSUE THREAD

**Issue by anrw** [dotansimha/graphql-code-generator#1442]
**Describe the bug**
When a type is implementing interface and fragments of different types are requested - codegen is generating a union of fragment types instead of intersection.

**To Reproduce**

Schema:
```javascript
scalar UUID

type File {
  title: String
  type: String
  url: String
}

interface HasAttachment {
  attachments: [File!]!
}

type Project implements HasAttachment {
  title: String,
  uuid: UUID!
  attachments: [File!]!
}

type Query {
  project(uuid: UUID!): Project
}
```
2. Documents:
```javascript
query findProject($projectUuid: UUID!) {
  project(uuid: $projectUuid) {
    uuid
    ...ProjectFields
    ...AttachmentsFields
  }
}

fragment ProjectFields on Project {
  title
}

fragment AttachmentsFields on HasAttachment {
  attachments {
    title
  }
}
```

3. Generated types:
![img](https://i.imgur.com/FMDORkn.png)

**Expected behavior**
```javascript
export type Project = {
  __typename?: "Project";
  uuid: Uuid;
} & (ProjectFields.Fragment & AttachmentsFields.Fragment);
```

**Environment:**
 - Codegen: 0.18.0

**Comment by dotansimha** [comment]
Hi @anrw ! 
Thank you for reporting this issue. 
We are currently working on a big refactor for TypeScript plugins, and we also fixed this issue. 
It's not correct that we always need to use `&`, because it cases where GraphQL interfaces or GraphQL unions are in user, and there are more fields, it should be both `|` and `&`. 
In the current refactor we added a code that handles it  (https://github.com/dotansimha/graphql-code-generator/blob/master/packages/plugins/visitor-plugin-common/src/selection-set-to-object.ts#L121-L131) by grouping the fragments by `type` and then using `&` to merge then, and `|` between them.

**Comment by brabeji** [comment]
Hello @dotansimha I'm running the latest `alpha` (kudos for publishing alpha commits!) I believe that in a more extreme case of fragments on union type, where united types share common interface, generated types are incorrect:

Schema:
```
schema {
  query: Query
}

type Query {
  widgets: [Widgets!]!
}

interface WidgetBase {
  id: ID!
}

type WidgetA implements WidgetBase {
  id: ID!
  size: Int!
}

type WidgetB implements WidgetBase {
  id: ID!
  color: String!
}

union Widgets = WidgetA | WidgetB
```

Documents:
```
query AQuery {
  widgets {
    ... on WidgetBase {
      id
    }
    ... on WidgetA {
      size
    }
    ... on WidgetB {
      color
    }
  }
}
```

Config:
```
generates:
  server-types.ts:
    - add: "/* tslint:disable */"
    - time
    - typescript
    - typescript-operations
```

I expect the resulting `AQueryQuery['widgets'][0]` type to have an `id` field.

Trying this in https://5c8bd8a4a550500008ff7d95--graphql-code-generator.netlify.com/ playground *and* locally with:
```
    "graphql-code-generator": "0.19.0-alpha.14bf4dca",
    "graphql-codegen-add": "0.19.0-alpha.14bf4dca",
    "graphql-codegen-fragment-matcher": "0.19.0-alpha.14bf4dca",
    "graphql-codegen-typescript": "0.19.0-alpha.2f3515aa",
    "graphql-codegen-typescript-graphql-files-modules": "0.19.0-alpha.14bf4dca",
    "graphql-codegen-typescript-operations": "0.19.0-alpha.2f3515aa",
```

Thanks!

**Comment by dotansimha** [comment]
Fix available in 1.0.0 👍
